src/IssueViz/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTagger.cs

```diff
@@ -40,11 +66,85 @@ protected override TagSpan<IntraTextAdornmentTag> CreateTagSpan(ISelectedIssueLo
             // should have zero length spans. Overriding this method allows control
             // over the tag spans.
             var adornmentSpan = new SnapshotSpan(trackedTag.Location.Span.Value.Start, 0);
-            var adornment = new IssueLocationAdornment(trackedTag, wpfView.FormattedLineSource);
+            var adornment = adornmentCache.CreateOrUpdate(trackedTag.Location);
 
-            // If we don't call Measure here the tag is positioned incorrectly
-            adornment.Measure(new System.Windows.Size(double.PositiveInfinity, double.PositiveInfinity));
             return new TagSpan<IntraTextAdornmentTag>(adornmentSpan, new IntraTextAdornmentTag(adornment, null, PositionAffinity.Predecessor));
         }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                tagAggregator.BatchedTagsChanged -= OnTagsChanged;
+            }
+            base.Dispose(disposing);
+        }
+
+        #region Adornment cache
+
+        internal interface ICachingAdornmentFactory
+        {
+            /// <summary>
+            /// Clears unnecessary entries from the cache
+            /// </summary>
+            /// <param name="currentLocations">The list of locations that are still in use</param>
+            void RemoveUnused(IEnumerable<IAnalysisIssueLocationVisualization> currentLocations);
```

Comment:
Maybe just "Refresh" or "RefreshCache"?
The convention is that "Remove" accepts the items to remove, so it's confusing that it's actually the items to keep.